File: src/main/scala/com/rwbymod/registry/CraftingRegistry.scala
package com.rwbymod.registry

import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraft.item.ItemStack
import scala.collection.JavaConversions._
import net.minecraft.init.Items
import net.minecraftforge.oredict.OreDictionary
import net.minecraftforge.oredict.ShapedOreRecipe
import com.rwbymod.dust.EnumDustType

object CraftingRegistry {

  def init {
    Array[(EnumDustType, EnumDustType, EnumDustType)](
      (EnumDustType.LIGHTNING, EnumDustType.WIND, EnumDustType.FIRE),
      (EnumDustType.ICE, EnumDustType.WIND, EnumDustType.WATER),
      (EnumDustType.STEAM, EnumDustType.FIRE, EnumDustType.WATER),
      (EnumDustType.GRAVITY, EnumDustType.EARTH, EnumDustType.WIND)
    ).foreach { ((out:EnumDustType, in1:EnumDustType, in2:EnumDustType) => {
      for(i <- 1 to 4) GameRegistry.addShapelessRecipe(
        new ItemStack(ItemRegistry.dust, i*2, out.ordinal),
        List.fill(i){ new ItemStack(ItemRegistry.dust, 1, in1.ordinal) } ++
        List.fill(i){ new ItemStack(ItemRegistry.dust, 1, in2.ordinal) }:_*
      )
    }).tupled }

    for(t <- EnumDustType.primary)
      GameRegistry.addSmelting(BlockRegistry.dustOre, new ItemStack(ItemRegistry.dust, 4, t.ordinal), .7f)

    for(t <- EnumDustType.all)
			GameRegistry.addShapedRecipe(
				new ItemStack(ItemRegistry.dustCrystal, 1, t.ordinal),
				" d ",
				"dqd",
				" d ",
				'd'.asInstanceOf[Character], new ItemStack(ItemRegistry.dust, 1, t.ordinal),
				'q'.asInstanceOf[Character], Items.QUARTZ
			)

		GameRegistry.addRecipe(new ShapedOreRecipe(
			new ItemStack(ItemRegistry.scroll),
			"ini",
			"ngn",
			"ici",
			new Character('i'), "ingotIron",
			'n'.asInstanceOf[Character], "nuggetIron",
			'g'.asInstanceOf[Character], "paneGlassColorless",
			'c'.asInstanceOf[Character], new ItemStack(ItemRegistry.dustCrystal, 1, OreDictionary.WILDCARD_VALUE)
		));
  }

}

File: src/old-java/com/rwbymod/misc/CraftingRegistry.java
package com.rwbymod.misc;

import com.rwbymod.block.BlockRegistry;
import com.rwbymod.item.ItemRegistry;

import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.oredict.OreDictionary;
import net.minecraftforge.oredict.ShapedOreRecipe;

public class CraftingRegistry {
	
	public static void init() {
		addDustMixingRecipe(EnumDustType.LIGHTNING, EnumDustType.WIND, EnumDustType.FIRE);
		addDustMixingRecipe(EnumDustType.ICE, EnumDustType.WIND, EnumDustType.WATER);
		addDustMixingRecipe(EnumDustType.STEAM, EnumDustType.WATER, EnumDustType.FIRE);
		addDustMixingRecipe(EnumDustType.GRAVITY, EnumDustType.EARTH, EnumDustType.WIND);
		
		for(EnumDustType type : EnumDustType.primary)
			GameRegistry.addSmelting(new ItemStack(BlockRegistry.dustOre, 1, type.ordinal()), new ItemStack(ItemRegistry.dust, 4, type.ordinal()), .7f);
		
		for(EnumDustType type : EnumDustType.all)
			GameRegistry.addShapedRecipe(
				new ItemStack(ItemRegistry.dustCrystal, 1, type.ordinal()),
				" d ",
				"dqd",
				" d ",
				'd', new ItemStack(ItemRegistry.dust, 1, type.ordinal()),
				'q', Items.QUARTZ
			);
		
		GameRegistry.addRecipe(new ShapedOreRecipe(
			new ItemStack(ItemRegistry.scroll), 
			"ini",
			"ngn",
			"ici",
			'i', "ingotIron",
			'n', "nuggetIron",
			'g', "paneGlassColorless",
			'c', new ItemStack(ItemRegistry.dustCrystal, 1, OreDictionary.WILDCARD_VALUE)
		));
	}
	
	private static void addDustMixingRecipe(EnumDustType out, EnumDustType in1, EnumDustType in2) {
		GameRegistry.addShapelessRecipe(
			new ItemStack(ItemRegistry.dust, 2, out.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal())
		);
		
		GameRegistry.addShapelessRecipe(
			new ItemStack(ItemRegistry.dust, 4, out.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal())
		);
		
		GameRegistry.addShapelessRecipe(
			new ItemStack(ItemRegistry.dust, 6, out.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal())
		);
		
		GameRegistry.addShapelessRecipe(
			new ItemStack(ItemRegistry.dust, 8, out.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in1.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal()),
			new ItemStack(ItemRegistry.dust, 1, in2.ordinal())
		);
	}

}
